File: app/crawler/puppeteer_service.py
"""Python wrapper for Node.js Puppeteer service"""
import logging
import httpx
from typing import List, Dict, Optional
from urllib.parse import urljoin, urlparse
from app.config import settings

logger = logging.getLogger(__name__)


class PuppeteerService:
    """Python wrapper for Node.js Puppeteer service"""

    def __init__(self, service_url: str = None):
        """
        Initialize Puppeteer service wrapper

        Args:
            service_url: URL of the Puppeteer service (defaults to config)
        """
        self.service_url = service_url or getattr(settings, 'PUPPETEER_SERVICE_URL', 'http://puppeteer-service:3000')
        self.base_url = f"{self.service_url.rstrip('/')}"

    async def crawl(self, company_name: str, career_url: str, timeout: int = 30000,
                    wait_for_selector: Optional[str] = None, wait_timeout: int = 30000) -> List[Dict]:
        """
        Request job extraction from Puppeteer service

        Args:
            company_name: Human-readable company name
            career_url: URL to company's career page
            timeout: Page load timeout in milliseconds
            wait_for_selector: CSS selector to wait for before extracting
            wait_timeout: Maximum time to wait for selector in milliseconds

        Returns:
            List of normalized job dictionaries
        """
        try:
            url = f"{self.base_url}/crawl"
            payload = {
                "company_name": company_name,
                "career_url": career_url,
                "timeout": timeout,
            }
            if wait_for_selector:
                payload["wait_for_selector"] = wait_for_selector
                payload["wait_timeout"] = wait_timeout

            async with httpx.AsyncClient(timeout=timeout / 1000 + 10) as client:
                logger.info(f"Requesting Puppeteer crawl for {company_name}: {career_url}")
                response = await client.post(url, json=payload)
                response.raise_for_status()

                result = response.json()
                if not result.get("success"):
                    logger.error(f"Puppeteer service returned error: {result.get('error')}")
                    return []

                jobs = result.get("jobs", [])
                logger.info(f"Puppeteer found {len(jobs)} jobs for {company_name}")

                # Normalize jobs
                normalized_jobs = []
                for job in jobs:
                    normalized = self._normalize_job(job, company_name, career_url)
                    if normalized:
                        normalized_jobs.append(normalized)

                return self._dedupe(normalized_jobs)

        except httpx.HTTPStatusError as e:
            logger.error(f"HTTP error from Puppeteer service: {e}")
            return []
        except httpx.TimeoutException:
            logger.error(f"Timeout waiting for Puppeteer service")
            return []
        except Exception as e:
            logger.error(f"Error calling Puppeteer service: {e}", exc_info=True)
            return []

    def _normalize_job(self, job_data: Dict, company_name: str, career_url: str) -> Optional[Dict]:
        """
        Normalize Puppeteer-extracted job data

        Args:
            job_data: Raw job data from Puppeteer service
            company_name: Company name
            career_url: Base career page URL

        Returns:
            Normalized job dictionary
        """
        try:
            title = job_data.get("title") or job_data.get("name") or job_data.get("position")
            if not title or not isinstance(title, str):
                return None
            title = title.strip()
            if not title:
                return None

            # Build job URL
            job_url = job_data.get("url") or job_data.get("href") or job_data.get("link")
            if job_url and not job_url.startswith("http"):
                job_url = urljoin(career_url, job_url)
            if not job_url:
                job_url = career_url

            # Extract location
            location = job_data.get("location") or job_data.get("city") or job_data.get("address")
            if isinstance(location, dict):
                location = location.get("name") or location.get("addressLocality") or str(location)

            # Generate external ID
            url_part = urlparse(job_url).path.replace("/", "_") or title.replace(" ", "_").lower()
            external_id = f"puppeteer_{company_name.lower().replace(' ', '_')}_{url_part}"[:255]

            normalized = {
                "external_id": external_id,
                "title": title,
                "company": company_name,
                "location": location.strip() if location else None,
                "url": job_url,
                "source_url": job_url,
                "description": job_data.get("description") or job_data.get("summary"),
                "job_type": job_data.get("job_type") or job_data.get("employmentType") or job_data.get("type"),
                "posted_date": None,  # Puppeteer extraction usually can't get reliable dates
                "platform": "puppeteer",
            }

            return normalized

        except Exception as e:
            logger.error(f"Error normalizing Puppeteer job: {e}", exc_info=True)
            return None

    def _dedupe(self, jobs: List[Dict]) -> List[Dict]:
        """Remove duplicate jobs"""
        seen = set()
        out: List[Dict] = []
        for j in jobs:
            key = (j.get('url') or '').lower().strip(), (j.get('title') or '').lower().strip()
            if key in seen or not key[0] or not key[1]:
                continue
            seen.add(key)
            out.append(j)
        return out

    async def health_check(self) -> bool:
        """
        Check if Puppeteer service is healthy

        Returns:
            True if service is healthy, False otherwise
        """
        try:
            url = f"{self.base_url}/health"
            async with httpx.AsyncClient(timeout=5.0) as client:
                response = await client.get(url)
                response.raise_for_status()
                return response.json().get("status") == "ok"
        except Exception as e:
            logger.debug(f"Puppeteer health check failed: {e}")
            return False

File: services/puppeteer-service.js
/**
 * Puppeteer Service - Node.js service for browser automation
 * Provides HTTP API for Python crawler to request browser-based job extraction
 */

const express = require('express');
const cors = require('cors');
const puppeteer = require('puppeteer');

const app = express();
const PORT = process.env.PORT || 3000;

// Middleware
app.use(cors());
app.use(express.json());

// Browser instance for reuse
let browserInstance = null;

/**
 * Initialize or get browser instance
 */
async function getBrowser() {
  if (!browserInstance) {
    browserInstance = await puppeteer.launch({
      headless: true,
      args: [
        '--no-sandbox',
        '--disable-setuid-sandbox',
        '--disable-dev-shm-usage',
        '--disable-accelerated-2d-canvas',
        '--disable-gpu'
      ]
    });
    console.log('Browser instance initialized');
  }
  return browserInstance;
}

/**
 * Health check endpoint
 */
app.get('/health', (req, res) => {
  res.json({ status: 'ok', service: 'puppeteer-service' });
});

/**
 * Crawl endpoint - Extract jobs from a career page
 * POST /crawl
 * Body: {
 *   company_name: string,
 *   career_url: string,
 *   timeout?: number (ms),
 *   wait_for_selector?: string,
 *   wait_timeout?: number (ms)
 * }
 */
app.post('/crawl', async (req, res) => {
  const { company_name, career_url, timeout = 30000, wait_for_selector, wait_timeout = 30000 } = req.body;

  if (!company_name || !career_url) {
    return res.status(400).json({ error: 'company_name and career_url are required' });
  }

  let browser = null;
  let page = null;

  try {
    // Get browser instance and create a new context
    const browserInstance = await getBrowser();
    browser = await browserInstance.createBrowserContext();

    // Create page
    page = await browser.newPage();
    await page.setViewport({ width: 1920, height: 1080 });
    await page.setUserAgent('Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36');

    console.log(`Crawling ${company_name}: ${career_url}`);

    // Navigate to page
    try {
      await page.goto(career_url, {
        waitUntil: 'networkidle',
        timeout: timeout
      });
    } catch (error) {
      console.warn(`Page load timeout for ${company_name}, continuing anyway`);
      // Continue even if timeout
    }

    // Wait for selector if provided
    if (wait_for_selector) {
      try {
        await page.waitForSelector(wait_for_selector, { timeout: wait_timeout });
        console.log(`Found selector ${wait_for_selector} for ${company_name}`);
      } catch (error) {
        console.warn(`Selector ${wait_for_selector} not found for ${company_name}`);
      }
    }

    // Wait for JavaScript to render
    await page.waitForTimeout(2000);

    // Extract jobs using multiple strategies
    const jobs = await extractJobs(page, company_name, career_url);

    res.json({
      success: true,
      jobs: jobs,
      count: jobs.length
    });

  } catch (error) {
    console.error(`Error crawling ${company_name}:`, error);
    res.status(500).json({
      success: false,
      error: error.message,
      jobs: []
    });
  } finally {
    // Cleanup
    if (page) {
      await page.close().catch(() => {});
    }
    if (browser) {
      await browser.close().catch(() => {});
    }
  }
});

/**
 * Extract jobs from page using multiple strategies
 */
async function extractJobs(page, companyName, careerUrl) {
  const jobs = [];

  // Strategy 1: Extract JSON-LD structured data
  try {
    const jsonLdJobs = await extractJsonLd(page);
    if (jsonLdJobs.length > 0) {
      jobs.push(...jsonLdJobs);
      console.log(`Found ${jsonLdJobs.length} jobs via JSON-LD for ${companyName}`);
    }
  } catch (error) {
    console.debug(`JSON-LD extraction failed: ${error.message}`);
  }

  // Strategy 2: Extract via common selectors
  if (jobs.length === 0) {
    try {
      const selectorJobs = await extractViaSelectors(page, careerUrl);
      if (selectorJobs.length > 0) {
        jobs.push(...selectorJobs);
        console.log(`Found ${selectorJobs.length} jobs via selectors for ${companyName}`);
      }
    } catch (error) {
      console.debug(`Selector extraction failed: ${error.message}`);
    }
  }

  // Strategy 3: Execute JavaScript to extract jobs
  if (jobs.length === 0) {
    try {
      const jsJobs = await extractViaJavaScript(page);
      if (jsJobs.length > 0) {
        jobs.push(...jsJobs);
        console.log(`Found ${jsJobs.length} jobs via JavaScript for ${companyName}`);
      }
    } catch (error) {
      console.debug(`JavaScript extraction failed: ${error.message}`);
    }
  }

  return dedupeJobs(jobs);
}

/**
 * Extract jobs from JSON-LD structured data
 */
async function extractJsonLd(page) {
  const jobs = [];
  try {
    const scripts = await page.$$eval('script[type="application/ld+json"]', (scripts) => {
      return scripts.map(script => script.textContent);
    });

    for (const content of scripts) {
      if (!content) continue;
      try {
        const data = JSON.parse(content);
        const jobPostings = parseJsonLdJobPosting(data);
        jobs.push(...jobPostings);
      } catch (e) {
        // Not valid JSON, skip
      }
    }
  } catch (error) {
    console.debug(`JSON-LD extraction error: ${error.message}`);
  }
  return jobs;
}

/**
 * Recursively parse JSON-LD for JobPosting objects
 */
function parseJsonLdJobPosting(data) {
  const jobs = [];
  if (typeof data !== 'object' || data === null) return jobs;

  if (Array.isArray(data)) {
    for (const item of data) {
      jobs.push(...parseJsonLdJobPosting(item));
    }
  } else {
    if (data['@type'] === 'JobPosting') {
      const jobLocation = data.jobLocation || {};
      const address = (typeof jobLocation === 'object' && jobLocation.address) || {};
      jobs.push({
        title: data.title || null,
        location: (typeof address === 'object' && address.addressLocality) || null,
        url: data.url || (data.hiringOrganization && data.hiringOrganization.sameAs) || null,
        job_type: data.employmentType || null,
        description: data.description || null
      });
    }
    // Recursively search nested objects
    for (const value of Object.values(data)) {
      if (typeof value === 'object' && value !== null) {
        jobs.push(...parseJsonLdJobPosting(value));
      }
    }
  }
  return jobs;
}

/**
 * Extract jobs using common CSS selectors
 */
async function extractViaSelectors(page, baseUrl) {
  const jobs = [];
  const selectors = [
    '[data-job-id]',
    '.job-listing',
    '.job-item',
    '.job-card',
    '[class*="job"]',
    '[id*="job"]',
    'article[class*="job"]',
    'div[class*="position"]',
    'li[class*="job"]'
  ];

  for (const selector of selectors) {
    try {
      const elements = await page.$$(selector);
      if (elements.length > 0) {
        console.debug(`Found ${elements.length} elements with selector ${selector}`);
        for (let i = 0; i < Math.min(elements.length, 50); i++) {
          try {
            const job = await extractJobFromElement(elements[i], baseUrl);
            if (job && job.title) {
              jobs.push(job);
            }
          } catch (error) {
            console.debug(`Error extracting job from element: ${error.message}`);
          }
        }
        if (jobs.length > 0) {
          break; // Found jobs, stop trying other selectors
        }
      }
    } catch (error) {
      // Selector not found, continue
    }
  }

  return jobs;
}

/**
 * Extract job data from a DOM element
 */
async function extractJobFromElement(element, baseUrl) {
  try {
    const title = await element.$eval('h2, h3, .title, [class*="title"]', el => el.textContent).catch(() => null);
    const link = await element.$('a[href]').catch(() => null);
    let url = null;
    if (link) {
      url = await link.evaluate(el => el.href).catch(() => null);
      if (url && !url.startsWith('http')) {
        url = new URL(url, baseUrl).href;
      }
    }
    const location = await element.$eval('[class*="location"], [class*="city"]', el => el.textContent).catch(() => null);

    if (title) {
      return {
        title: title.trim(),
        url: url,
        location: location ? location.trim() : null
      };
    }
  } catch (error) {
    // Element doesn't have expected structure
  }
  return null;
}

/**
 * Extract jobs by executing JavaScript on the page
 */
async function extractViaJavaScript(page) {
  const jobs = [];
  try {
    const result = await page.evaluate(() => {
      const jobs = [];
      // Look for common job data patterns in JavaScript
      const scripts = Array.from(document.querySelectorAll('script'));
      for (const script of scripts) {
        const text = script.textContent || script.innerHTML;
        if (text && (text.includes('jobs') || text.includes('positions') || text.includes('openings'))) {
          try {
            // Try to extract JSON from script
            const jsonMatch = text.match(/\{[\s\S]*"jobs"[\s\S]*\}/);
            if (jsonMatch) {
              const data = JSON.parse(jsonMatch[0]);
              if (data.jobs && Array.isArray(data.jobs)) {
                return data.jobs;
              }
            }
          } catch (e) {
            // Not valid JSON
          }
        }
      }
      // Look for window/global job data
      if (window.jobs && Array.isArray(window.jobs)) {
        return window.jobs;
      }
      if (window.__INITIAL_STATE__ && window.__INITIAL_STATE__.jobs) {
        return window.__INITIAL_STATE__.jobs;
      }
      return [];
    });

    if (Array.isArray(result)) {
      jobs.push(...result.filter(item => item && typeof item === 'object'));
    }
  } catch (error) {
    console.debug(`JavaScript extraction error: ${error.message}`);
  }
  return jobs;
}

/**
 * Remove duplicate jobs
 */
function dedupeJobs(jobs) {
  const seen = new Set();
  const unique = [];
  for (const job of jobs) {
    const key = `${(job.url || '').toLowerCase().trim()}|${(job.title || '').toLowerCase().trim()}`;
    if (key && !seen.has(key) && job.title) {
      seen.add(key);
      unique.push(job);
    }
  }
  return unique;
}

// Start server
async function startServer() {
  try {
    // Initialize browser on startup (optional - can be lazy)
    // await getBrowser();
    app.listen(PORT, '0.0.0.0', () => {
      console.log(`Puppeteer service listening on port ${PORT}`);
    });
  } catch (error) {
    console.error('Failed to start server:', error);
    process.exit(1);
  }
}

// Graceful shutdown
process.on('SIGTERM', async () => {
  console.log('SIGTERM received, shutting down gracefully');
  if (browserInstance) {
    await browserInstance.close();
  }
  process.exit(0);
});

process.on('SIGINT', async () => {
  console.log('SIGINT received, shutting down gracefully');
  if (browserInstance) {
    await browserInstance.close();
  }
  process.exit(0);
});

startServer();

